pipx/pipxrc.py

```diff
@@ -0,0 +1,196 @@
+        return self._val_or_default(self.pipxrc_info.venv_metadata, default)
+
+    def get_injected_packages(
+        self, default: List[Dict[str, Any]]
+    ) -> List[Dict[str, Any]]:
+        if self.pipxrc_info.injected_packages is not None:
+            injected_packages = []
+            for package in self.pipxrc_info.injected_packages:
+                package_info = {"package": package}
+                package_info.update(
+                    self.pipxrc_info.injected_packages[package]._asdict()
+                )
+                injected_packages.append(package_info)
+            return injected_packages
+        else:
+            return default
+
+    def set_package_or_url(self, package_or_url: str) -> None:
+        # TODO 20190923: if package_or_url is a local path, we need to make it
+        #   an absolute path
+        self.pipxrc_info.package_or_url = package_or_url
+
+    def set_venv_metadata(self, venv_metadata: PipxVenvMetadata) -> None:
+        self.pipxrc_info.venv_metadata = venv_metadata
+
+    def set_install_options(
+        self, pip_args: List[str], venv_args: List[str], include_dependencies: bool
+    ) -> None:
+        self.pipxrc_info.install = InstallOptions(
+            pip_args=pip_args,
+            venv_args=venv_args,
+            include_dependencies=include_dependencies,
+        )
+
+    def add_injected_package(
+        self,
+        package: str,
+        pip_args: List[str],
+        verbose: bool,
+        include_apps: bool,
+        include_dependencies: bool,
+        force: bool,
+    ) -> None:
+        if self.pipxrc_info.injected_packages is None:
+            self.pipxrc_info.injected_packages = {}
+
+        self.pipxrc_info.injected_packages[package] = InjectedPackage(
+            pip_args=pip_args,
+            verbose=verbose,
+            include_apps=include_apps,
+            include_dependencies=include_dependencies,
+            force=force,
+        )
+
+    def write(self) -> None:
+        # If writing out, make sure injected_packages is not None, so next
+        #   successful read of pipxrc does not use default in
+        #   get_injected_packages()
+        if self.pipxrc_info.injected_packages is None:
+            self.pipxrc_info.injected_packages = {}
+
+        # TODO 20190919: raise exception on failure?
+        with open(self.venv_dir / PIPX_INFO_FILENAME, "w") as pipxrc_fh:
+            json.dump(
+                self.pipxrc_info.to_dict(),
+                pipxrc_fh,
+                indent=4,
+                sort_keys=True,
+                cls=JsonEncoderPipx,
+            )
+
+    def read(self) -> None:
+        try:
+            with open(self.venv_dir / PIPX_INFO_FILENAME, "r") as pipxrc_fh:
+                self.pipxrc_info.from_dict(
+                    json.load(pipxrc_fh, object_hook=_json_decoder_object_hook)
+                )
+        except IOError:  # Reset self.pipxrc_info if problem reading
```

Comment:
suggest logging a message like "Failed to read pipxrc file at {path}"

Reply:
Sounds good.

Reply:
Done. With a `logging.warning`